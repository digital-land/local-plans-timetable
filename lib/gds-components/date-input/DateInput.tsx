/**
 * GDS component: https://design-system.service.gov.uk/components/date-input/
 */
import cn from "classnames";

export interface DateInputProps {
  label?: string;
  name: string;
  onChange: (value: string) => void;
  value: string;
  id: string;
  error?: string;
  withDay?: boolean;
}

export const DateInput = ({
  value,
  label,
  name,
  onChange,
  error,
  id,
  withDay = false,
}: DateInputProps) => {
  const [year = "", month = "", day = ""] = value.split("-");

  const handleChange = (value: string) => {
    onChange(value.replace(/-/g, "").length ? value : "");
  };

  return (
    <div
      className={cn("govuk-form-group", { "govuk-form-group--error": error })}
      id={id}
    >
      <fieldset
        className="govuk-fieldset"
        role="group"
        aria-describedby={`${name}-hint`}
      >
        {label && (
          <legend className="govuk-fieldset__legend govuk-fieldset__legend--m">
            {label}
          </legend>
        )}
        {error && (
          <p className="govuk-error-message">
            <span className="govuk-visually-hidden">Error:</span> {error}
          </p>
        )}
        <div id={`${name}-hint`} className="govuk-hint">
          For example,{withDay ? " 12" : ""} 8 2024
        </div>
        <div className="govuk-date-input">
          {withDay && (
            <div className="govuk-date-input__item">
              <div className="govuk-form-group">
                <label
                  className="govuk-label govuk-date-input__label"
                  htmlFor={`${name}-day`}
                >
                  Day
                </label>
                <input
                  id={`${name}-day`}
                  data-testid={`${name}-day`}
                  className="govuk-input govuk-date-input__input govuk-input--width-2"
                  type="text"
                  inputMode="numeric"
                  value={day}
                  onChange={(e) =>
                    handleChange(`${year}-${month}-${e.target.value}`)
                  }
                  onBlur={(e) =>
                    e.target.value.length == 1 &&
                    handleChange(`${year}-${month}-${"0" + e.target.value}`)
                  }
                  maxLength={2}
                />
              </div>
            </div>
          )}
          <div className="govuk-date-input__item">
            <div className="govuk-form-group">
              <label
                className="govuk-label govuk-date-input__label"
                htmlFor={`${name}-month`}
              >
                Month
              </label>
              <input
                id={`${name}-month`}
                data-testid={`${name}-month`}
                className="govuk-input govuk-date-input__input govuk-input--width-2"
                type="text"
                inputMode="numeric"
                value={month}
                onChange={(e) =>
                  handleChange(
                    `${year}-${e.target.value}${withDay ? `-${day}` : ""}`
                  )
                }
                onBlur={(e) =>
                  e.target.value.length == 1 &&
                  handleChange(
                    `${year}-${"0" + e.target.value}${withDay ? `-${day}` : ""}`
                  )
                }
                maxLength={2}
              />
            </div>
          </div>
          <div className="govuk-date-input__item">
            <div className="govuk-form-group">
              <label
                className="govuk-label govuk-date-input__label"
                htmlFor={`${name}-year`}
              >
                Year
              </label>
              <input
                id={`${name}-year`}
                data-testid={`${name}-year`}
                className="govuk-input govuk-date-input__inpaut govuk-input--width-4"
                type="text"
                inputMode="numeric"
                value={year}
                onChange={(e) =>
                  handleChange(
                    `${e.target.value}-${month}${withDay ? `-${day}` : ""}`
                  )
                }
                maxLength={4}
              />
            </div>
          </div>
        </div>
      </fieldset>
    </div>
  );
};
