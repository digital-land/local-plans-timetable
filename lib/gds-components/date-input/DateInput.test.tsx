import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { useState } from "react";

import { DateInput, DateInputProps } from "./DateInput";

interface DateInputWrapperProps extends Pick<DateInputProps, "onChange" | "withDay"> {
  initialValue: string;
}

const DateInputWrapper = ({
  initialValue,
  onChange,
  withDay
}: DateInputWrapperProps) => {
  const [value, setValue] = useState(initialValue);

  return (
    <DateInput
      label="Date"
      name="date"
      value={value}
      withDay={withDay}
      onChange={(value) => {
        setValue(value);
        onChange(value);
      }}
    />
  );
};

it("shows the right input values from the provided value", () => {
  const { asFragment } = render(
    <DateInputWrapper initialValue="2007-03" onChange={() => {}} />
  );

  expect(asFragment()).toMatchSnapshot();
  expect(screen.getByLabelText("Month")).toHaveValue("03");
  expect(screen.getByLabelText("Year")).toHaveValue("2007");
});

it("shows input for day when withDay is true", () => {
  const { asFragment } = render(
    <DateInputWrapper initialValue="2007-03-01" onChange={() => {}} withDay />
  );

  expect(asFragment()).toMatchSnapshot();
  expect(screen.getByLabelText("Day")).toHaveValue("01");
  expect(screen.getByLabelText("Month")).toHaveValue("03");
  expect(screen.getByLabelText("Year")).toHaveValue("2007");
});

it("calls the onChange handler with the right value when the month input changes", async () => {
  const onChange = jest.fn();
  const { asFragment } = render(
    <DateInputWrapper initialValue="2007-03" onChange={onChange} />
  );

  await userEvent.clear(screen.getByLabelText("Month"));
  await userEvent.type(screen.getByLabelText("Month"), "4");
  await userEvent.tab()

  expect(asFragment()).toMatchSnapshot();
  expect(onChange).toHaveBeenLastCalledWith("2007-04");
});

it("calls the onChange handler with the right value when the year input changes", async () => {
  const onChange = jest.fn();
  const { asFragment } = render(
    <DateInputWrapper initialValue="2007-03" onChange={onChange} />
  );

  await userEvent.clear(screen.getByLabelText("Year"));
  await userEvent.type(screen.getByLabelText("Year"), "2008");

  expect(asFragment()).toMatchSnapshot();
  expect(onChange).toHaveBeenLastCalledWith("2008-03");
});
