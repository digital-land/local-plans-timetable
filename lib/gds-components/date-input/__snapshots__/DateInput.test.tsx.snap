// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calls the onChange handler with the right value when the month input changes 1`] = `
<DocumentFragment>
  <div
    class="govuk-form-group"
  >
    <fieldset
      aria-describedby="date-hint"
      class="govuk-fieldset"
      role="group"
    >
      <legend
        class="govuk-fieldset__legend govuk-fieldset__legend--m"
      >
        Date
      </legend>
      <div
        class="govuk-hint"
        id="date-hint"
      >
        For example, 8 2024
      </div>
      <div
        class="govuk-date-input"
      >
        <div
          class="govuk-date-input__item"
        >
          <div
            class="govuk-form-group"
          >
            <label
              class="govuk-label govuk-date-input__label"
              for="date-month"
            >
              Month
            </label>
            <input
              class="govuk-input govuk-date-input__input govuk-input--width-2"
              data-testid="date-month"
              id="date-month"
              inputmode="numeric"
              maxlength="2"
              type="text"
              value="04"
            />
          </div>
        </div>
        <div
          class="govuk-date-input__item"
        >
          <div
            class="govuk-form-group"
          >
            <label
              class="govuk-label govuk-date-input__label"
              for="date-year"
            >
              Year
            </label>
            <input
              class="govuk-input govuk-date-input__inpaut govuk-input--width-4"
              data-testid="date-year"
              id="date-year"
              inputmode="numeric"
              maxlength="4"
              type="text"
              value="2007"
            />
          </div>
        </div>
      </div>
    </fieldset>
  </div>
</DocumentFragment>
`;

exports[`calls the onChange handler with the right value when the year input changes 1`] = `
<DocumentFragment>
  <div
    class="govuk-form-group"
  >
    <fieldset
      aria-describedby="date-hint"
      class="govuk-fieldset"
      role="group"
    >
      <legend
        class="govuk-fieldset__legend govuk-fieldset__legend--m"
      >
        Date
      </legend>
      <div
        class="govuk-hint"
        id="date-hint"
      >
        For example, 8 2024
      </div>
      <div
        class="govuk-date-input"
      >
        <div
          class="govuk-date-input__item"
        >
          <div
            class="govuk-form-group"
          >
            <label
              class="govuk-label govuk-date-input__label"
              for="date-month"
            >
              Month
            </label>
            <input
              class="govuk-input govuk-date-input__input govuk-input--width-2"
              data-testid="date-month"
              id="date-month"
              inputmode="numeric"
              maxlength="2"
              type="text"
              value="03"
            />
          </div>
        </div>
        <div
          class="govuk-date-input__item"
        >
          <div
            class="govuk-form-group"
          >
            <label
              class="govuk-label govuk-date-input__label"
              for="date-year"
            >
              Year
            </label>
            <input
              class="govuk-input govuk-date-input__inpaut govuk-input--width-4"
              data-testid="date-year"
              id="date-year"
              inputmode="numeric"
              maxlength="4"
              type="text"
              value="2008"
            />
          </div>
        </div>
      </div>
    </fieldset>
  </div>
</DocumentFragment>
`;

exports[`shows input for day when withDay is true 1`] = `
<DocumentFragment>
  <div
    class="govuk-form-group"
  >
    <fieldset
      aria-describedby="date-hint"
      class="govuk-fieldset"
      role="group"
    >
      <legend
        class="govuk-fieldset__legend govuk-fieldset__legend--m"
      >
        Date
      </legend>
      <div
        class="govuk-hint"
        id="date-hint"
      >
        For example, 12 8 2024
      </div>
      <div
        class="govuk-date-input"
      >
        <div
          class="govuk-date-input__item"
        >
          <div
            class="govuk-form-group"
          >
            <label
              class="govuk-label govuk-date-input__label"
              for="date-day"
            >
              Day
            </label>
            <input
              class="govuk-input govuk-date-input__input govuk-input--width-2"
              data-testid="date-day"
              id="date-day"
              inputmode="numeric"
              maxlength="2"
              type="text"
              value="01"
            />
          </div>
        </div>
        <div
          class="govuk-date-input__item"
        >
          <div
            class="govuk-form-group"
          >
            <label
              class="govuk-label govuk-date-input__label"
              for="date-month"
            >
              Month
            </label>
            <input
              class="govuk-input govuk-date-input__input govuk-input--width-2"
              data-testid="date-month"
              id="date-month"
              inputmode="numeric"
              maxlength="2"
              type="text"
              value="03"
            />
          </div>
        </div>
        <div
          class="govuk-date-input__item"
        >
          <div
            class="govuk-form-group"
          >
            <label
              class="govuk-label govuk-date-input__label"
              for="date-year"
            >
              Year
            </label>
            <input
              class="govuk-input govuk-date-input__inpaut govuk-input--width-4"
              data-testid="date-year"
              id="date-year"
              inputmode="numeric"
              maxlength="4"
              type="text"
              value="2007"
            />
          </div>
        </div>
      </div>
    </fieldset>
  </div>
</DocumentFragment>
`;

exports[`shows the right input values from the provided value 1`] = `
<DocumentFragment>
  <div
    class="govuk-form-group"
  >
    <fieldset
      aria-describedby="date-hint"
      class="govuk-fieldset"
      role="group"
    >
      <legend
        class="govuk-fieldset__legend govuk-fieldset__legend--m"
      >
        Date
      </legend>
      <div
        class="govuk-hint"
        id="date-hint"
      >
        For example, 8 2024
      </div>
      <div
        class="govuk-date-input"
      >
        <div
          class="govuk-date-input__item"
        >
          <div
            class="govuk-form-group"
          >
            <label
              class="govuk-label govuk-date-input__label"
              for="date-month"
            >
              Month
            </label>
            <input
              class="govuk-input govuk-date-input__input govuk-input--width-2"
              data-testid="date-month"
              id="date-month"
              inputmode="numeric"
              maxlength="2"
              type="text"
              value="03"
            />
          </div>
        </div>
        <div
          class="govuk-date-input__item"
        >
          <div
            class="govuk-form-group"
          >
            <label
              class="govuk-label govuk-date-input__label"
              for="date-year"
            >
              Year
            </label>
            <input
              class="govuk-input govuk-date-input__inpaut govuk-input--width-4"
              data-testid="date-year"
              id="date-year"
              inputmode="numeric"
              maxlength="4"
              type="text"
              value="2007"
            />
          </div>
        </div>
      </div>
    </fieldset>
  </div>
</DocumentFragment>
`;
